lib/java/src/org/apache/thrift/async/TAsyncMethodCall.java

```diff
@@ -225,8 +231,13 @@ private void cleanUpAndFireCallback(SelectionKey key) {
     key.interestOps(0);
     // this ensures that the TAsyncMethod instance doesn't hang around
     key.attach(null);
```

Comment:
AFAIK, key is not removed from `selector.keys()` unless `key.cancel()` is called, so `interestOps(0)` and `attach(null)` is not enough (maybe wrong assumption ?).
Do you happen to have any idea why the original code does not call `cancel()` for the success path here ?

As your intention here seems to be to prevent `key.cancel()` in `transition()` for the error path, I thought you might know about it.